### PR TITLE
Allow SSO authentication to provide a user secret

### LINKED ISSUE #24838: Support for per-user encrypted files in SSO setting.
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

**Is your feature request related to a problem? Please describe.**
Currently the default encryption plugin cannot have user-keys with single sign on, such as SAML. As described [here](https://docs.nextcloud.com/server/20/admin_manual/configuration_files/encryption_details.html#key-type-user-key) the problem is that NC does not have access to the password, that is used to secure the users private key.

**Describe the solution you'd like**
In our setting, our SAML identity provider can already safely store and provide per-user secrets derived from the users password.
Therefore the obvious solution seems to to let the authentication backend provide a stable per-user secret which can be used in lieu of the actual password.

**Describe alternatives you've considered**
A more flexible alternative would be to let the keystore use a generic and configurable "user secret provider". Different authentication backends could support different strategies for providing such a secret. However, my knowledge of the Nextcloud  codebase is not good enough to implement this.

**Additional context**
I submitted two PR which implement this strategy.

* First [we need to fetch the secret in the server](https://github.com/nextcloud/server/pull/24837) from the user backend.
* Second in user_saml [the secret must be received from the idp](https://github.com/nextcloud/user_saml/pull/498) and made available.

I am not quite sure TBH what the approach is for changing an interface. There might be other implementations of the `Authentication\IApacheBackend` interface. And also the change must be synchronized between server and saml plugin.

We are currently running this patch in our testing environment. We can successfully create users, login via browsers and app, and normally use all features. The configuration has encryption enabled and the master-key disabled. We confirmed that files are correctly encrypted and decrypted.

### PR DESCRIPTION
**Pull request by MichaIng**
Implementing PR #24837 from immerda

As of my failed attempt to fix #27909 😄.

### CODE REVIEW

File: lib/public/Authentication/IApacheBackend.php [hunk 1]
```diff
@@ -56,4 +56,12 @@ public function getLogoutUrl();
 	 * @since 6.0.0
 	 */
 	public function getCurrentUserId();
+
+	/**
+	 * Optionally returns a stable per-user secret. This secret is for
+	 * instance used to secure file encryption keys.
+	 * @return string|null
+	 * @since 21.0.0
```
Comment (juliushaertl) [inline lib/public/Authentication/IApacheBackend.php:64]:
```suggestion
	 * @since 23.0.0
```

Comment (MichaIng) [reply lib/public/Authentication/IApacheBackend.php:64]: Or shall we backport this back to NC21? It doesn't look like it could cause any issues.

Comment (juliushaertl) [reply lib/public/Authentication/IApacheBackend.php:64]: The interface change will break any existing implementation that doesn't have this method so it would be a breaking change.

Comment (MichaIng) [reply lib/public/Authentication/IApacheBackend.php]: Okay, done.

**Comment by summersab** [comment]
Thanks, @MichaIng! Sorry for the hassle.

Do you think you could give me a hand with figuring out why this PR is failing? It's the companion code for the `user_saml` app:
https://github.com/nextcloud/user_saml/pull/537

File: lib/public/Authentication/IApacheBackend.php [hunk 2]
```diff
@@ -56,4 +56,12 @@ public function getLogoutUrl();
...
 	 */
 	public function getCurrentUserId();
+
+	/**
+	 * Optionally returns a stable per-user secret. This secret is for
+	 * instance used to secure file encryption keys.
+	 * @return string|null
+	 * @since 23.0.0
+	 */
+	public function getCurrentUserSecret();
```
Comment (juliushaertl) [inline lib/public/Authentication/IApacheBackend.php:66]: For having this in a separate interface as mentioned in https://github.com/nextcloud/server/pull/24837#discussion_r674125504 you would need to create a new file next to this one with just the interface definition and the new method.

Besides that the autoloader needs to be updated with running `./build/autoloaderchecker.sh` and committing the results of the run as well.

Comment (MichaIng) [reply lib/public/Authentication/IApacheBackend.php:66]: Okay, I hope I did it right 😅.

Comment (MichaIng) [reply lib/public/Authentication/IApacheBackend.php:66]: Ah, forgot to remove the function here.

File: lib/private/legacy/OC_User.php [hunk 3]
```diff
@@ -172,7 +172,9 @@ public static function loginWithApache(\OCP\Authentication\IApacheBackend $backe
 				$userSession = \OC::$server->getUserSession();
 				$userSession->setLoginName($uid);
 				$request = OC::$server->getRequest();
-				$userSession->createSessionToken($request, $uid, $uid);
+				$secret = $backend->getCurrentUserSecret();
+				$userSession->createSessionToken($request, $uid, $uid, $secret);
+				$password = $secret === null ? '' : $secret;
```
Comment (juliushaertl) [inline lib/private/legacy/OC_User.php:177]:
```suggestion
				$password = null;
				if ($backend instanceof IProvideUserSecretBackend) {
					$password = $backend->getCurrentUserSecret();
				}
				$userSession->createSessionToken($request, $uid, $uid, $password);
```

File: lib/private/legacy/OC_User.php [hunk 4]
```diff
@@ -184,7 +186,7 @@ public static function loginWithApache(\OCP\Authentication\IApacheBackend $backe
 					'post_login',
 					[
 						'uid' => $uid,
-						'password' => '',
+						'password' => $password,
```
Comment (juliushaertl) [inline lib/private/legacy/OC_User.php:189]:
```suggestion
						'password' => $password ?? '',
```

File: lib/public/Authentication/IProvideUserSecretBackend.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+interface IProvideUserSecretBackend {
+
+	/**
+	 * Optionally returns a stable per-user secret. This secret is for
+	 * instance used to secure file encryption keys.
+	 * @return string
+	 * @since 23.0.0
+	 */
+	public function getCurrentUserSecret();
```
Comment (juliushaertl) [inline lib/public/Authentication/IProvideUserSecretBackend.php:40]:
```suggestion
	public function getCurrentUserSecret(): string;
```

Comment (juliushaertl) [reply lib/public/Authentication/IProvideUserSecretBackend.php:40]: Basically if the interface is implemented we assume then that there is always a user secret returned.

File: lib/private/legacy/OC_User.php (hunk 3 above)
Comment (MichaIng) [reply lib/private/legacy/OC_User.php]: Suggestions applied and rebased. But there are two issues, I think:
- The function argument is defined as
    ```php
    \OCP\Authentication\IApacheBackend $backend
    ```
    So I think we need to load the new backend additionally?
- And `IProvideUserSecretBackend` => `OCP\Authentication\IProvideUserSecretBackend` using the namespace?

Comment (juliushaertl) [reply lib/private/legacy/OC_User.php]: The argument type is fine since the `IApacheBackend` would always need to be implemented by the backend, `IProvideUserSecretBackend` is basically an optional extension then.

However for the second part you will need to either use the full namespace or add a `use OCP\Authentication\IApacheBackend` to the top list of the file.

**Comment by MichaIng** [comment]
CI is happy now.

File: lib/private/legacy/OC_User.php [hunk 6]
```diff
@@ -172,7 +172,11 @@ public static function loginWithApache(\OCP\Authentication\IApacheBackend $backe
 				$userSession = \OC::$server->getUserSession();
 				$userSession->setLoginName($uid);
 				$request = OC::$server->getRequest();
-				$userSession->createSessionToken($request, $uid, $uid);
+				$password = null;
+				if ($backend instanceof OCP\Authentication\IProvideUserSecretBackend) {
```
Comment (juliushaertl) [inline lib/private/legacy/OC_User.php:176]:
```suggestion
				if ($backend instanceof \OCP\Authentication\IProvideUserSecretBackend) {
```

Needs a trailing backslash as otherwise it would be considered to be within the current namespace, otherwise good from my side :+1:

File: lib/public/Authentication/IProvideUserSecretBackend.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ */
+interface IProvideUserSecretBackend {
+
+	/**
+	 * Optionally returns a stable per-user secret. This secret is for
+	 * instance used to secure file encryption keys.
+	 * @return string
+	 * @since 23.0.0
+	 */
+	public function getCurrentUserSecret(): string;
```
Comment (LukasReschke) [inline lib/public/Authentication/IProvideUserSecretBackend.php:40]: Would making this nullable make sense? Especially as `$password = null;` is the default.
```suggestion
	public function getCurrentUserSecret(): ?string;
```

**Review by LukasReschke — APPROVED** [review]

File: lib/public/Authentication/IProvideUserSecretBackend.php (hunk 7 above)
Comment (MichaIng) [reply lib/public/Authentication/IProvideUserSecretBackend.php:40]: I think the idea was to explicitly not allow a null value here, so that checks can be omitted when the interface is implemented. Backends then need to always provide a string.

Ref: https://github.com/nextcloud/server/pull/27929#discussion_r684146422

But I see this is inconsistent with the description "Optionally returns a stable per-user secret.", which just has not been changed when I copy&pasted it from `IApacheBackend` (where it needed to be optional, of course) to its own interface. Shall I replace with:
```
* Returns a stable per-user secret e.g. used to secure file encryption keys.
```

Comment (summersab) [reply lib/public/Authentication/IProvideUserSecretBackend.php:40]: Two comments/issues:

1. I've tried the most recent version of this PR with the `IProvideUserSecretBackend` in conjunction with the associated user_saml plugin PR:
https://github.com/nextcloud/user_saml/pull/537
 .
Since users logging in with this method would be using the `user_saml` backend, `lib/private/legacy/OC_User.php:177` never gets executed. Therefore, the password is never set. Am I missing something?
2. With the code from the original PR, there seems to be a problem. After a few minutes, `heartbeat` requests start returning 401 responses. Then, any `getstoragestats` request triggers a page refresh after a 401 response because the session is no longer valid. I have yet to determine what is causing the session to fail, but it is a problem.

Any thoughts? I have a local Keycloak -> NC dev setup working, and I'm happy to test.

**Comment by MichaIng** [comment]
@summersab 
Since your post has nothing to do with the reviewed line/suggestion, I'll answer here 🙂: If I understood it right, the `user_saml` backend now needs to provide both, `IApacheBackend` for the base functionality and `IProvideUserSecretBackend` additionally for the secret: https://github.com/nextcloud/user_saml/blob/d4f2d9b/lib/UserBackend.php#L40

**Comment by summersab** [comment]
@MichaIng - apologies, I figured I'd reply to the most recent thread so that everyone would take notice. I didn't want broken code to get approved.

I applied your suggested change, but the second issue I mentioned still persists - `getstoragestats` requests return a 401 and trigger a page refresh. I'm not sure what is causing the session to expire, but I will try to step through the code and find the problem.

**Comment by MichaIng** [comment]
> apologies

Not problem, I just want to avoid that your issue gets hidden when the question about nullable return gets resolved. Everyone who is subscribed to this topic should get a notification, regardless whether the post was done within a review block or here.

> `getstoragestats` requests return a 401 and trigger a page refresh

Hmm, does this work when using the current master branch versions of backend, without user secret then, of course? This doesn't sound like something related to this change but a change in how/when a session gets invalidated or who to keep it valid.

Let me rebase onto current master, just in case something even was fixed in this regards.

**Comment by summersab** [comment]
A few updates. First, without passing a password, the invalid session issue doesn't happen. I tried debugging the situation WITH a password set, but I don't know enough about how the session handing works to figure it out.

It is worth mentioning that when the password IS set and a request (such as `getstoragestats`) triggers a refresh, the browser redirects back to the IdP, but the IdP session and cookies are still valid. The IdP immediately redirects the user back to NC and is logged in automatically. So, the user is never logged out of the IdP and forced to log back in manually - they just get temporarily redirected to the IdP and immediately logged back into NC. It's . . . weird.

Lastly, with regards to this discussion:
https://github.com/nextcloud/server/pull/27929#discussion_r695632608

Without a password, `getCurrentUserSecret()` will return null. So, this suggested change should either be reversed (probably the recommended solution) or a try/catch should be added here:
https://github.com/nextcloud/server/blob/4940ac90d95d56fc0a3e1f5fb9413af8b08d5c53/lib/private/legacy/OC_User.php#L177

File: lib/public/Authentication/IProvideUserSecretBackend.php (hunk 7 above)
Comment (MichaIng) [reply lib/public/Authentication/IProvideUserSecretBackend.php:40]: > Without a password, `getCurrentUserSecret()` will return null. So, this suggested change should either be reversed (probably the recommended solution) or a try/catch should be added

Is it possible to "provide" the `IProvideUserSecretBackend` only when a user secret is actually set? In theory this would then reduce some overhead, when the backend is used without it and would make allowing/checking for `null` obsolete, right? But I clearly lack knowledge about theses things, so probably it's nonsense 😄.

**Comment by summersab** [comment]
@MichaIng I discovered that the Passwords app may be a good place to look for testing. It performs an auth check before loading, and with the SAML password set, it throws an exception every time. I'm still trying to trace what is happening, but maybe you can take a look.

Also, I notice that regular authentication (and SAML auth without this password mod) uses the `oc_authtoken` table to store tokens. However, the `oc_user_saml_auth_token` table is never used by the user_saml app. Perhaps there needs to be a modification made to the app to utilize this table when a password value is set?

**Comment by MichaIng** [comment]
> Perhaps there needs to be a modification made to the app to utilize this table when a password value is set?

This is also going above my knowledge of the Nextcloud code and how it deals with such things, what it does automatically and what is this the backends/apps task. I'm currently a bit busy on our project but I'll have a look into the passwords app when I find time. Could you paste the exception here as a starting point?

**Comment by summersab** [comment]
I'm going to tag @immerda to see if maybe they can weigh in, but I'll keep digging. I may send them an email if I don't get anywhere.

**Comment by summersab** [comment]
I believe I found it. I submitted an issue to the `user_saml` project here in order to have a discussion prior to opening a PR:
https://github.com/nextcloud/user_saml/issues/547

My suggested solution is at the bottom of the ticket.

**Comment by immerda** [comment]
I can confirm what many people on this PR feared. When we set the password to `null` the token will never be considered valid and the user is not actually logged in.

To be consistent with the previous implementation I think it would be good to return an empty string instead. Also note that if anybody has per-user encryption enabled with sso, then the password is currently the empty string and changing anything there will make the decryption key unreadable.

File: lib/private/legacy/OC_User.php [hunk 8]
```diff
@@ -190,7 +194,7 @@ public static function loginWithApache(\OCP\Authentication\IApacheBackend $backe
 					'post_login',
 					[
 						'uid' => $uid,
-						'password' => null,
+						'password' => $password ?? null,
```
Comment (immerda) [inline lib/private/legacy/OC_User.php:197]: isn't that the same as just `$password`

Comment (summersab) [reply lib/private/legacy/OC_User.php:197]: @immerda exactly - this was discussed in a section now marked "resolved" here: https://github.com/nextcloud/server/pull/27929#discussion_r684145691

From what I can tell (and since I have been using this code almost daily for months, now), if the password is set to an empty string instead of null, that fixes the problem. That should never have been changed from your original PR.

This has been pushed back to NC 24, but I suppose that's best considering this is being discussed so late in the game.

**Comment by immerda** [comment]
@MichaIng @summersab I dug into the code surrounding all of this and also verified with live instances in various configurations (with/without SSO, with/without encryption). Here are some conclusions what I understand so far:

* There are passwordless authtokens: if the password is set to null, then nextcloud knows that the token does not have a pw. In this case the checkPassword is not called.
* Therefore the correct thing to do, when there is no userSecret from the user backend is to pass `null` to `createSessionToken`.
* The user_saml app has a completely bogus `checkPassword` implementation. It checks tokens from a table (user_saml_auth_token) which is *never* populated.  This `checkPassword` function is essentially identical to a function that is just `return false`. This bug has not surfaced, since checkPassword is never called for passwordless tokens.
* The current user_saml sso-secret patch is broken: the `UserBackend` does not implement the new `IProvideUserSecretBackend` interface. That is why `$backend instanceof \OCP\Authentication\IProvideUserSecretBackend` is always false and nothing happens here.

In summary for this PR when the sso secret is not provided, the pw should be set to null. To achieve this there are two ways forward for this PR:

1. Leave as is, and change the user_saml patch to conditionally implement the `IProvideUserSecretBackend` interface
2. Change the type of the `getCurrentUserSecret` interface to `?string`. This allows the user backend to indicate that there is no user secret provided, by just returning `null`.

I strongly lean towards (2) as option (1) would make the patch to user_saml quite a bit more complicated. We would need some sort of factory method for instantiating the correct UserBackend depending on if a `user_secret_mapping` configuration variable is set.

File: lib/private/legacy/OC_User.php (hunk 8 above)
Comment (immerda) [reply lib/private/legacy/OC_User.php:197]: ah, sorry, I am starting to get really lost here! My comment was just that passing `$password` alone, without `?? null` should be functionally equivalent. I was not arguing that we should pass an empty string.

But now I see there are two different things:

1. to `createSessionToken` we must pass `null` for it to create a passwordless token.
2. to `OC_Hook::emit` we might have to pass an empty string, however I am not sure. I am running the patch against NC 21 and it seems to work either way...

File: lib/public/Authentication/IProvideUserSecretBackend.php (hunk 7 above)
Comment (immerda) [reply lib/public/Authentication/IProvideUserSecretBackend.php:40]: I have a proposal to get us unstuck here. How about we make `IProvideUserSecretBackend::getCurrentUserSecret()` nullable, but then do the following:

```
$password = null;
if ($backend instanceof \OCP\Authentication\IProvideUserSecretBackend) {
  $password = $backend->getCurrentUserSecret();
}
if ($password !== null) {
  $userSession->createSessionToken($request, $uid, $uid, $password);
} else {
  $userSession->createSessionToken($request, $uid, $uid);
}
```

This would allow us to disentangle the two problems, preserve the identical behavior for the non-userSecret case and let `createSessionToken` decide what to do when there is no password. This solution would also be more robust, as it removes that dependency on the decision if token passwords are nullable or not and allows us to define an interface that is independent of that.

File: lib/private/legacy/OC_User.php (hunk 8 above)
Comment (MichaIng) [reply lib/private/legacy/OC_User.php:197]: Indeed `$password ?? null` is nonsense as `??` applies only if `$password` is `null` already 😄.

File: lib/public/Authentication/IProvideUserSecretBackend.php (hunk 7 above)
Comment (MichaIng) [reply lib/public/Authentication/IProvideUserSecretBackend.php:40]: Isn't it exactly the same outcome whether passing `$password = null` or no argument at all? The `createSessionToken` has it's fourth argument `null` in both cases, isn't it?

The main question is indeed whether we want to be more explicit and force usage of a password when the backend is implemented or whether we want to allow implementing the backend without the need to provide a password. Generally I'm fans of being more explicit, but I understand that it may make the backend implementation more complicated.

I think it is agreed that we can or even should merge this PR first, before doing so at the backend, especially since the interfaces are defined here, so the above is really the last open question. I think we need someone who is more in charge to simply do a decision and we'll be able to deal with the result. But we need to get unstuck :smilley:.

**Comment by MichaIng** [comment]
Rebased and resolved conflict. We need someone to decide whether the API function should be allowed to return null or not (see discussion in code review).

**Comment by immerda** [comment]
Two things happened:

1. There is a discussion in this PR if our new interface should be allowed to return null. But everybody in this PR is in favor of returning non-null values. As the original author I agree and it will be possible to rewrite the sso backend for this new api, by simply implementing the behavior in a subclass of the [UserBackend](https://github.com/nextcloud/user_saml/blob/master/lib/UserBackend.php#L41). This should be easy, as there is only one [central location](https://github.com/nextcloud/user_saml/blob/d58d532a680eff21120f60a57062e475a6dd313d/appinfo/app.php#L50) where it is instantiated. 
2. There was a discussion in wider NC if the password that we pass to createSessionToken can be null. This is indeed a discussion that afaik has not finished. But we resolved the issue here by not passing any more nulls to that function than the code already did. so our PR is decoupled from that discussion.

In summary, the people behind this PR agree on the api, the original author agrees, the api works for the intended usecase and this PR is not blocked by any external decision.

So I think we took that decision that you are talking about. If not, please let me know who can take this decision.

**Comment by MichaIng** [comment]
So we just need another approval.

The failing S3 checks are unrelated, fail randomly, I'll try to resolve with a rebase before merging.

**Review by ChristophWurst — APPROVED** [review]
Looks good
